### DISCUSSION TITLE
Can't click on a button added to a layer

### DISCUSSION THREAD

**Discussion by aceinetx** [axmolengine/axmol#2277] [Q&A] [answered]
Just switched from cocos2dx to axmol (for android reasons), and I can't click on a button that I added to a layer, in a scene it works fine (The callback doesn't get called, the button texture doesn't change)
Here's the way I do this in cocos2dx:
```cpp
auto button = Button::create("CloseNormal.png", "CloseSelected.png");

button->setPosition(Vec2(240, 160));

button->addClickEventListener([this](Ref *sender) { 
  CCLOG("Hi!");
});

this->addChild(button);
```
Basically the same thing in axmol:
```cpp
bool MenuLayer::init()
{
  if (!Layer::init())
    return false;

  screen_size = Director::getInstance()->getWinSize();
  auto label  = Label::createWithTTF("Test", "fonts/arial.ttf", 20.5f);
  label->setPosition({screen_size.x / 2, screen_size.y - 10});

  auto button = Button::create("CloseNormal.png", "CloseSelected.png");

  button->setPosition(Vec2(240, 160));

  button->addClickEventListener([this](Object* sender) {
    AXLOG("Hi!");
  });

  this->addChild(button);
  this->addChild(label);

  return true;
}
```
Here's how the button looks in cocos explorer:
![image](https://github.com/user-attachments/assets/8d5eba8b-1261-4989-a475-2edc5acb5852)

I also tried using menus, but it didn't change anything

**Comment by rh101** [comment]
Which version of Axmol? 
Which platform did you test it on?

**Reply by aceinetx** [reply]
> Which version of Axmol? Which platform did you test it on?

v2.2.1, Windows

**Comment by rh101** [comment]
Can you provide the entire `MenuLayer` file? Also, how is the instance of `MenuLayer` created, and what options in it are configured?

**Reply by aceinetx** [reply]
> Can you provide the entire `MenuLayer` file? Also, how is the instance of `MenuLayer` created, and what options in it are configured?

> What is `MenuLayer`? A Scene? Layer? Something else?
> 
> Can you provide that entire file that sets up MenuLayer, with the button code etc.? The reason is that the parent object may be blocking input to the child nodes, but there is no way to tell unless we see how you're setting it up.

MenuLayer.h
```cpp
#if !defined(MENU_LAYER_H)
#define MENU_LAYER_H

#include "axmol.h"
#include "ui/axmol-ui.h"
#include "PlayLayer.h"
#include <iostream>

class MenuLayer : public ax::Layer
{
private:
    ax::Vec2 screen_size;

public:
    bool init() override;

    void onEnter() override;

    CREATE_FUNC(MenuLayer);
};

#endif  // MENU_LAYER_H
```

MenuLayer.cpp:
```cpp
#include "MenuLayer.h"

using namespace ax;
using namespace ax::ui;

bool MenuLayer::init()
{
    if (!Layer::init())
        return false;

    screen_size = Director::getInstance()->getWinSize();
    auto label  = Label::createWithTTF("Test", "fonts/arial.ttf", 20.5f);
    label->setPosition({screen_size.x / 2, screen_size.y - 10});


    auto button = Button::create("CloseNormal.png", "CloseSelected.png");

    button->setPosition(Vec2(240, 160));

    button->addClickEventListener([this](Object* sender) { AXLOG("Hi!");
    });

    this->addChild(button);

    this->addChild(label);

    return true;
}

void MenuLayer::onEnter() {
    AXLOG("Parent: %p", this->getParent());
}
```

MainScene.h:
```cpp
#ifndef __MAIN_SCENE_H__
#define __MAIN_SCENE_H__

#include "axmol.h"
#include "ui/axmol-ui.h"

#include "MenuLayer.h"
#include "CocosExplorer.h"

class MainScene : public ax::Scene
{
public:
    bool init() override;

    void menuCloseCallback(ax::Object* sender);
};

#endif  // __MAIN_SCENE_H__
```

MainScene.cpp:
```cpp
#include "MainScene.h"

using namespace ax;
using namespace ax::ui;

bool MainScene::init()
{
    if (!Scene::init())
    {
        return false;
    }


    auto menu = MenuLayer::create();
    this->addChild(menu);

    CocosExplorer::openForever();

    scheduleUpdate();

    return true;
}

void MainScene::menuCloseCallback(ax::Object* sender)
{
    _director->end();
}

```

**Reply by rh101** [reply, marked as answer]
The mistake is here:
```
void MenuLayer::onEnter() {
    AXLOG("Parent: %p", this->getParent());
}
```

It should be:
```
void MenuLayer::onEnter() {
    Layer::onEnter();
    AXLOG("Parent: %p", this->getParent());
}
```

The same goes for any other methods that are overriden; don't forget to call the base class implementation unless there is a specific reason not to do so.

You can generally assume that functionality related to core engine features, like user input and such, would not have obvious issues like this, at least in common/typical cases of usage, so you can safely assume that the source of the problem would most likely be in the project code.

**Reply by aceinetx** [reply]
> The mistake is here:
> 
> ```
> void MenuLayer::onEnter() {
>     AXLOG("Parent: %p", this->getParent());
> }
> ```
> 
> It should be:
> 
> ```
> void MenuLayer::onEnter() {
>     Layer::onEnter();
>     AXLOG("Parent: %p", this->getParent());
> }
> ```
> 
> The same goes for any other methods that are overriden; don't forget to call the base class implementation unless there is a specific reason not to do so.

Oh god can't believe I was this dumb, thank you for help!
